src/System.Management.Automation/engine/CommandCompletion/CompletionHelpers.cs

```diff
@@ -110,5 +108,40 @@ internal static bool CompletionRequiresQuotes(string completion, bool escape)
 
         private static bool ContainsCharsToCheck(ReadOnlySpan<char> text, bool escape)
             => text.ContainsAny(escape ? s_escapeCharsToCheck : s_defaultCharsToCheck);
+
+        /// <summary>
+        /// Quotes and escapes a given completion text based on the specified options.
+        /// </summary>
+        /// <param name="completionText">
+        /// The text to be quoted and potentially escaped.
+        /// </param>
+        /// <param name="quote">
+        /// The quote character to use for enclosing the text. Defaults to a single quote if not provided.
+        /// </param>
+        /// <param name="escapeSingleQuoteChars">
+        /// Indicates whether single quote characters in the text should be escaped. Defaults to <c>true</c>.
+        /// </param>
+        /// <returns>
+        /// The quoted and optionally escaped version of the <paramref name="completionText"/>.
+        /// </returns>
+        internal static string QuoteCompletionText(
+            string completionText,
+            string quote,
+            bool escapeSingleQuoteChars = true)
+        {
+            if (!CompletionRequiresQuotes(completionText, false))
```

Comment:
I wonder if its better using `ContainsCharactersRequiringQuotes`: https://github.com/PowerShell/PowerShell/blob/3d2a3e106ed0e9f5e69e9b981e97f1d20456b6fb/src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs#L5184-L5187

Atleast we can escape all chars  with single quotes instead just `$` and backticks

Reply:
As a repetition of my comment from another PR. I'd prefer to get such improvements in separate PSs with full description and new tests for covering scenarios.

Reply:
@iSazonov Yep fair enough. I think I am not confident making such a change anyways without more investigation.

I will freeze any new changes on this PR except for review feedback. I think current changes alone are good enough to refactor a lot of code.